sqlglot/lineage.py

```diff
@@ -276,8 +282,66 @@ def to_node(
                 reference_node_name=reference_node_name,
                 trim_selects=trim_selects,
             )
+
+        elif pivot:
+            if isinstance(pivot.args.get("columns"), list):
+                # The source is a pivot operation, so we need to trace back to the aggregated column
+                pivot_column_mapping = {}
+                pivot_aggs_count = len(pivot.expressions)
+                columns = pivot.args["columns"]
+                columns_count = len(columns)
+                column_name = c.name
+                downstream_columns = []
+
+                if any(column_name == pivoted_column.name for pivoted_column in columns):
+                    # The column is in the pivot, so we need to trace back to the aggregation that created it.
+                    for i, agg in enumerate(pivot.expressions):
+                        agg_cols = list(agg.find_all(exp.Column))
+                        for col_index in range(i, columns_count, pivot_aggs_count):
+                            # e.g. "pivot (sum(value) as value_sum, max(price)) for category in ('a' as cat_a, 'b')"
+                            # For each aggregation function, the pivot creates a new column for each field in category combined with the aggfunc.
+                            # So the columns parsed have this order: cat_a_value_sum, cat_a, b_value_sum, b.
+                            # Because of this step wise manner the aggfunc 'sum(value) as value_sum' belongs to the column indices 0, 2,
+                            # and the aggfunc 'max(price)' without an alias belongs to the column indices 1, 3.
+                            # Here only the columns used in the aggregations are of interest in the lineage, so lookup the pivot column name
+                            # by index and map that with the columns used in the aggregation.
+                            pivot_column_mapping[columns[col_index].name] = agg_cols
```

Comment:
Most of this code looks loop-invariant. The only thing that blocks us from factoring it out and computing it once is `pivot` AFAICT, because we use `c.table` in its definition. 

You could do:

```python
        pivot = next((p for p in scope.pivots if p.alias_or_name and not p.unpivot), None)
```

And then compute `pivot_column_mapping` once and cache it so it can be reused in the following iterations. Possibly worth creating a helper function to do that so this section isn't as "polluted" with boilerplate.

Note that you'd also update the branch condition to make it `pivot and pivot.alias_or_name == c.table`.

Reply:
I think so, this assumption is also made at the `qualify` level, so if a query is not fully-qualified (which will probably be the case for queries with pivots), this won't work anyway.